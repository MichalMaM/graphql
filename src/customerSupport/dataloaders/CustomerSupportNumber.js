// @flow

import DataLoader from 'dataloader';
import idx from 'idx';
import { get } from '../../common/services/HttpRequest';

export default function createCustomerSupportNumbersLoader() {
  return new DataLoader((countryLocales: $ReadOnlyArray<string>) =>
    batchLoad(countryLocales),
  );
}

async function batchLoad(
  countryLocales: $ReadOnlyArray<string>,
): Promise<Array<?Object>> {
  const customerSupportNumbers = await fetchCustomerSupportNumbers();
  return countryLocales.map(countryLocale => {
    const customerSupportNumber =
      customerSupportNumbers[countryLocale] || customerSupportNumbers.default;
    return customerSupportNumber !== undefined ? customerSupportNumber : null;
  });
}

async function fetchCustomerSupportNumbers() {
  const response = await get(
    'https://booking-api.skypicker.com/api/v0.1/configs/kiwicom',
  );
  const enabled = idx(response, _ => _.contacts.phones.enabled);

  if (enabled) {
    return idx(response, _ => _.contacts.phones.locales) || {};
  }
  return {};
}
