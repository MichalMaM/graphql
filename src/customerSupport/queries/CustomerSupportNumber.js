// @flow

import GraphQLCustomerSupportNumber from '../types/outputs/CustomerSupportNumber';
import type { GraphqlContextType } from '../../common/services/GraphqlContext';

export default {
  type: GraphQLCustomerSupportNumber,
  description: 'Get customer support number for your locale',
  resolve: async (
    ancestor: mixed,
    args: Object,
    { dataLoader, locale }: GraphqlContextType,
  ) => {
    const customerSupportNumber = await dataLoader.customerSupportNumber.load(
      locale.territory.toLocaleLowerCase(),
    );

    return customerSupportNumber;
  },
};
