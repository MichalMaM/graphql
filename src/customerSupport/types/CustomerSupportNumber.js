// @flow

export type CustomerSupportNumber = {|
  +id: string,
  +number: string,
|};
