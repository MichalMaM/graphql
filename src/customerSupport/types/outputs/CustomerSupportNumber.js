// @flow

import { GraphQLObjectType, GraphQLString } from 'graphql';

import type { CustomerSupportNumber } from '../CustomerSupportNumber';

export default new GraphQLObjectType({
  name: 'CustomerSupportNumber',
  fields: {
    localeTerritory: {
      type: GraphQLString,
      description: 'Customer Support phone number locale territory',
      resolve: ({ id }: CustomerSupportNumber): string => id,
    },

    number: {
      type: GraphQLString,
      description: 'Customer Support phone number',
      resolve: ({ number }: CustomerSupportNumber): string => number,
    },
  },
});
