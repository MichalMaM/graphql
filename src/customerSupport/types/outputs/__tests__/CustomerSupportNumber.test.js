// @flow

import CustomerSupportNumber from '../CustomerSupportNumber';

it('CustomerSupportNumber type should have valid fields', () => {
  expect(CustomerSupportNumber.getFields()).toMatchSnapshot();
});
