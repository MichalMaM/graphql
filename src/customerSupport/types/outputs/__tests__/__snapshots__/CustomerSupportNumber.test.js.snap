// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerSupportNumber type should have valid fields 1`] = `
Object {
  "localeTerritory": Object {
    "args": Array [],
    "description": "Customer Support phone number locale territory",
    "isDeprecated": false,
    "name": "localeTerritory",
    "resolve": [Function],
    "type": "String",
  },
  "number": Object {
    "args": Array [],
    "description": "Customer Support phone number",
    "isDeprecated": false,
    "name": "number",
    "resolve": [Function],
    "type": "String",
  },
}
`;
